### PR TITLE
tso: add GetLastTS to get last global TSO

### PR DESCRIPTION
**Pull request by Andrewmatilde**
Signed-off-by: Andrewmatilde <davis6813585853062@outlook.com>

### What problem does this PR solve?
ref [#21522](https://github.com/pingcap/tidb/issues/21522)

### Check List

Tests

- Unit test

Side effects

- Increased code complexity

### Release note

 No release note

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -638,11 +642,26 @@ func (c *client) GetTS(ctx context.Context) (physical int64, logical int64, err
...
+
 func (c *client) GetLocalTS(ctx context.Context, dcLocation string) (physical int64, logical int64, err error) {
 	resp := c.GetLocalTSAsync(ctx, dcLocation)
 	return resp.Wait()
 }
 
+func (c *client) GetLastLocalTS(ctx context.Context, dcLocation string) (int64, int64, error) {
+	lastTSOInterface, ok := c.lastTSMap.Load(dcLocation)
+	if !ok {
+		err := fmt.Errorf("func GetLastLocalTS fail : get last local TS before get first TS or something wrong happen")
```
Comment (JmPotato) [inline client/client.go:657]: Please make the log info look like this:

```shell
[pd] get last local TS before get first TS or something wrong happen
```

File: client/client.go [hunk 2]
```diff
@@ -56,6 +56,10 @@ type Client interface {
 	GetLocalTS(ctx context.Context, dcLocation string) (int64, int64, error)
 	// GetLocalTSAsync gets a local timestamp from PD, without block the caller.
 	GetLocalTSAsync(ctx context.Context, dcLocation string) TSFuture
+	// GetLastTS gets the last timestamp given by GetTS.
+	GetLastTS(ctx context.Context) (int64, int64, error)
+	// GetLastLocalTS gets the last timestamp given by GetLastLocalTS.
+	GetLastLocalTS(ctx context.Context, dcLocation string) (int64, int64, error)
```
Comment (JmPotato) [inline client/client.go:62]: Maybe we can merge these two methods into a single one?

File: client/client.go (hunk 1 above)
Comment (rleungx) [reply client/client.go:657]: How about using errors.New?

Comment (Andrewmatilde) [reply client/client.go:657]: thanks for your review, I agree with you.

File: client/client.go [hunk 3]
```diff
@@ -638,11 +642,26 @@ func (c *client) GetTS(ctx context.Context) (physical int64, logical int64, err
...
+	return c.GetLastLocalTS(ctx, globalDCLocation)
+}
+
 func (c *client) GetLocalTS(ctx context.Context, dcLocation string) (physical int64, logical int64, err error) {
 	resp := c.GetLocalTSAsync(ctx, dcLocation)
 	return resp.Wait()
 }
 
+func (c *client) GetLastLocalTS(ctx context.Context, dcLocation string) (int64, int64, error) {
+	lastTSOInterface, ok := c.lastTSMap.Load(dcLocation)
```
Comment (Yisaer) [inline client/client.go:655]: According to https://github.com/pingcap/tidb/issues/21522 , we also need to maintain the arrival timestamp in lastTSMap.

**Review by Yisaer — COMMENTED** [review]
Rest LGTM

**Comment by Yisaer** [comment]
```sh
revive -formatter friendly -config revive.toml $(go list ./...)
  ✘  https://revive.run/r#confusing-results  unnamed results of the same type may be confusing, consider using named results  
  /home/runner/work/pd/pd/client/client.go:654:1
```

plz solve this error

File: client/client.go (hunk 2 above)
Comment (Andrewmatilde) [reply client/client.go]: Thanks for your review, I agree with you.

**Comment by nolouch** [comment]
Why need this interface?

File: client/client.go (hunk 3 above)
Comment (Andrewmatilde) [reply client/client.go]: Thanks for your review, I agree with you.

**Comment by Yisaer** [comment]
> https://github.com/pingcap/tidb/blob/master/store/tikv/oracle/oracles/pd.go#L134-L152

I miss that, but I think If we implement this function in pdclient, other components could reuse this ability?

**Comment by nolouch** [comment]
@Yisaer  It's not a common request. and there is an issue about how to push the last ts if not have any request, TiDB already solve it.

**Comment by Yisaer** [comment]
> @Yisaer It's not a common request. and there is an issue about how to push forward the `lastTS` if not have any request, TiDB already solves it . BTW, the `GetLastTS` is ambiguous, whether it means the client's `lastTS` or the server's `lastTS`, the interface also confuse me.

got it. That's ok to me.

**Pull request closed without merging**